### PR TITLE
Reading Error response properly in AndroidClientHandler 

### PR DESCRIPTION
**Pull request by smstuebe**
- Removed java flame comments.
- try to read `ErrorStream` if error occurred

> **Response Handling**
> HttpURLConnection will follow up to five HTTP redirects. It will follow redirects from one origin server to another. This implementation doesn't follow redirects from HTTPS to HTTP or vice versa.
If the HTTP response indicates that an error occurred, getInputStream() will throw an IOException. Use getErrorStream() to read the error response. 

https://developer.android.com/reference/java/net/HttpURLConnection.html#getErrorStream()

Sorry for not updating / writing test. Could not build the test project.

### CODE REVIEW

**Comment by jonpryor** [comment]
@grendello?

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 1]
```diff
@@ -267,11 +267,8 @@ protected void AssertSelf ()
 					// HttpClientHandler throws an exception in this instance, but I think it's not a good
 					// idea. We'll return the response message with all the information required by the
 					// application to fill in the blanks and provide the requested credentials instead.
-					//
-					// We should return the body of the response too but, alas, the Java client will throw
-					// a, wait for it, FileNotFound exception if we attempt to access the input stream. So
-					// no body, just a dummy. Java FTW!
-					ret.Content = new StringContent ("Unauthorized", Encoding.ASCII);
```
Comment (grendello) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: There's no reason to remove the entire section of the comment. If you think that wording here is indeed inflammatory, please rephrase as there's actual information here.

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 2]
```diff
@@ -267,11 +267,8 @@ protected void AssertSelf ()
...
+
+					ret.Content = GetErrorContent(httpConnection,  new StringContent ("Unauthorized", Encoding.ASCII));
```
Comment (grendello) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Formatting - please follow the [Mono Coding Guidelines](http://www.mono-project.com/community/contributing/coding-guidelines/#style-guidelines), in this case there should be a space following the `GetErrorContent` call.

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 3]
```diff
@@ -285,28 +282,18 @@ protected void AssertSelf ()
 					ret.RequestedAuthentication = RequestedAuthentication;
 					return ret;
 			}
-
+
```
Comment (grendello) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Please remove whitespace-only changes

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 4]
```diff
@@ -285,28 +282,18 @@ protected void AssertSelf ()
...
-					if (encodings != null) {
-						if (encodings.Contains (GZIP_ENCODING, StringComparer.OrdinalIgnoreCase))
-							inputStream = new GZipStream (inputStream, CompressionMode.Decompress);
-						else if (encodings.Contains (DEFLATE_ENCODING, StringComparer.OrdinalIgnoreCase))
-							inputStream = new DeflateStream (inputStream, CompressionMode.Decompress);
-					}
-				}
-				ret.Content = new StreamContent (inputStream);
-			} else {
+				ret.Content = GetContent(httpConnection, httpConnection.InputStream);
```
Comment (grendello) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Formatting (method call space)

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 5]
```diff
@@ -285,28 +282,18 @@ protected void AssertSelf ()
...
+			}
+			else {
 				if (Logger.LogNet)
-					Logger.Log (LogLevel.Info, LOG_APP, $"Status code is {statusCode}, returning empty content");
-				// For 400 >= response code <= 599 the Java client throws the FileNotFound exeption when attempting to read from the connection
-				// Client tests require we return no content here
```
Comment (grendello) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Please restore the comment in appropriate place, it contains actual useful information for the coder

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 6]
```diff
@@ -333,6 +320,32 @@ protected void AssertSelf ()
 				Logger.Log (LogLevel.Info, LOG_APP, $"Returning");
 			return ret;
 		}
+		
+		private HttpContent GetErrorContent (HttpURLConnection httpConnection, HttpContent fallbackContent)
```
Comment (grendello) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: `private` is implied and [not necessary here](http://www.mono-project.com/community/contributing/coding-guidelines/#baroque-coding), please remove

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 7]
```diff
@@ -333,6 +320,32 @@ protected void AssertSelf ()
 				Logger.Log (LogLevel.Info, LOG_APP, $"Returning");
 			return ret;
 		}
+		
+		private HttpContent GetErrorContent (HttpURLConnection httpConnection, HttpContent fallbackContent) 
+		{
+			var contentStream = httpConnection.ErrorStream;
+
+			if (contentStream != null) {
+				return GetContent(httpConnection, contentStream);
```
Comment (grendello) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Formatting (method call space)

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 8]
```diff
@@ -333,6 +320,32 @@ protected void AssertSelf ()
...
+			}
+
+			return fallbackContent;
+		}
+
+	    private HttpContent GetContent (URLConnection httpConnection, Stream contentStream)
```
Comment (grendello) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: ditto re `private`

File: src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs [hunk 9]
```diff
@@ -333,6 +320,32 @@ protected void AssertSelf ()
...
+                string[] encodings = httpConnection.ContentEncoding?.Split(',');
+                if (encodings != null) {
+                    if (encodings.Contains(GZIP_ENCODING, StringComparer.OrdinalIgnoreCase))
+                        inputStream = new GZipStream(inputStream, CompressionMode.Decompress);
+                    else if (encodings.Contains(DEFLATE_ENCODING, StringComparer.OrdinalIgnoreCase))
+                        inputStream = new DeflateStream(inputStream, CompressionMode.Decompress);
+                }
+            }
+            return new StreamContent(inputStream);
+        }
```
Comment (grendello) [inline src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs]: Please reformat the entire method according to the coding guidelines

**Comment by grendello** [comment]
Commented inline

**Comment by smstuebe** [comment]
done. Better?

**Comment by jonpryor** [comment]
build

**Comment by jonpryor** [comment]
build

**Comment by martinwiboe** [comment]
This problem seems to still occur on the latest Stable release of Xamarin.Android (7.0.1.2). When we read the `Content` property, an empty `StringContent` is returned. Are the changes in this PR part of the latest stable release?

**Comment by roubachof** [comment]
@martinwiboe I can see the commit in the cycle 9 branch: https://github.com/xamarin/xamarin-android/blob/cycle9/src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs

It's not in the cycle8-sr0, but it could be in cycle8 SR1 or 2, but we can only speculate since there are no branches for these releases.

**Comment by djdmbrwsk** [comment]
I can confirm that this fix is in Xamarin.Android 7.1.0.30.  Which is (at least currently) available in the Beta channel.

**Pull request merged**